arangod/Pregel/Algos/AIR/Accumulators.cpp

```diff
@@ -0,0 +1,300 @@
+  : Accumulator<VPackSlice>(options, defs) {
+  _definition = defs.at(*options.customType);
+  greenspun::InitMachine(_machine);
+  SetupFunctions();
+  if (options.parameters) {
+    _parameters = *options.parameters;
+  }
+}
+
+auto CustomAccumulator<VPackSlice>::clear() -> greenspun::EvalResult {
+  VPackBuilder result;
+  return greenspun::Evaluate(_machine, _definition.clearProgram.slice(), result).mapError([](auto& err) {
+    err.wrapMessage("in clearProgram of custom accumulator");
+  });
+}
+
+auto CustomAccumulator<VPackSlice>::setBySlice(VPackSlice v)
+  -> arangodb::greenspun::EvalResult {
+  inputSlice = v;
+  TRI_DEFER({ inputSlice = Slice::noneSlice(); })
+
+    if (_definition.setProgram.isEmpty()) {
+      _buffer.clear();
+      _buffer.add(v);
+      _value = _buffer.slice();
+      return {};
+    }
+
+  VPackBuilder result;
+  return greenspun::Evaluate(_machine, this->_definition.setProgram.slice(), result)
+    .mapError([](auto& err) {
+      err.wrapMessage("in setProgram of custom accumulator");
+    });
+}
+
+auto CustomAccumulator<VPackSlice>::getIntoBuilder(VPackBuilder& result)
+  -> arangodb::greenspun::EvalResult {
+  if (_definition.getProgram.isEmpty()) {
+    result.add(_value);
+    return {};
+  }
+
+  return greenspun::Evaluate(_machine, this->_definition.getProgram.slice(), result)
+    .mapError([](auto& err) {
+      err.wrapMessage("in getProgram of custom accumulator");
+    });
+}
+
+// FIXME: duplicate code here and below
+auto CustomAccumulator<VPackSlice>::updateByMessageSlice(VPackSlice msg) -> greenspun::EvalResultT<UpdateResult> {
+  this->inputSlice = msg.get("value");
+  this->inputSender =  msg.get("sender");
+
+  TRI_DEFER({
+      this->inputSlice = VPackSlice::noneSlice();
+      this->inputSender = VPackSlice::noneSlice(); })
+
+    VPackBuilder result;
+  auto res = greenspun::Evaluate(_machine, _definition.updateProgram.slice(), result);
+  if (res.fail()) {
+    return res.error().wrapMessage("in updateProgram of custom accumulator");
+  }
+
+  auto resultSlice = result.slice();
+  if (resultSlice.isString()) {
+    if (resultSlice.isEqualString("hot")) {
+      return UpdateResult::CHANGED;
+    } else if (resultSlice.isEqualString("cold")) {
+      return UpdateResult::NO_CHANGE;
+    }
+  } else if (resultSlice.isNone()) {
+    return AccumulatorBase::UpdateResult::NO_CHANGE;
+  }
+  return greenspun::EvalError(
+    "update program did not return a valid value: expected `\"hot\"` or "
+    "`\"cold\"`, found: " +
+    result.toJson());
+}
+
+auto CustomAccumulator<VPackSlice>::updateByMessage(MessageData const& msg)
+    -> greenspun::EvalResultT<UpdateResult> {
+  this->inputSlice = msg._value.slice();
+  VPackBuilder sender;
+  sender.add(VPackValue(msg._sender));
+  this->inputSender = sender.slice();
+
+  TRI_DEFER({
+    this->inputSlice = VPackSlice::noneSlice();
+    this->inputSender = VPackSlice::noneSlice();
+  })
+
+  VPackBuilder result;
```

Comment:
This was a fast (and not so cheap) way of getting something working. We already have some ideas how to improve the whole AIR VM. Currently it is just working with VPackSlices and such. Later we would like to replace this with a better datatype that uses less memory, is easier to work with and faster because of less copying stuff around.

Reply:
Ok